minetest/minetest
# Faster blit_with_alpha()

The `blit_with_alpha` function has a noticeable effect on the time it takes to join a game.

To reduce the join times, I replace the `blit_with_alpha` function with a new one:
* It does not uses floating-point numbers.
* It directly operates on the raw pixel data instead of using the comparatively
  slow `setPixel` and `getPixel` functions from Irrlicht.
  Only ECF_A8R8G8B8 base images are supported now.
  If the top image does not have the ECF_A8R8G8B8 colour format, it is converted;
  I assume that this happens rarely.
* There are case distinctions for fully opaque, fully transparent and semi-transparent pixels.
  This empirically increases the performance since the mixing between two semi-transparent happens rarely.
* The new function no longer has the `src_pos` argument since it was always the zero vector.
* The function is only documented once where it is declared.

For backwards compatibility, `blit_with_alpha` still mixes colours without gamma correction.
`blit_with_alpha` nonetheless behaves slightly different than before:
If a semi-transparent pixel is drawn on top of another semi-transparent pixel,
the colour is mixed in a way which we can consider to be more correct now.

Related issue: #14322
Roadmap goal: 2.1 Rendering/Graphics improvements (perhaps)

## To do

This PR is a Ready for Review.



## How to test

* Add this Lua code to a mod:
```Lua
local kick_on_join = true
if kick_on_join then
	minetest.register_on_joinplayer(function(player)
		minetest.kick_player(player:get_player_name(),
			"This is for testing purposes. Don't cry.")
	end)
end
```
* Start a local Minetest server
* In a Shell which supports `time`, join the server with the following command:
```sh
time bin/minetest --name your_name --address 127.0.0.1 --go
```
* Repeat the above command multiple times with and without the proposed changes and note down the duration values.

With the mods I commonly use, I could observe a speedup from 2.5 s to 2.2 s real time and a speedup from 1.3 s to 1.0 s user CPU time.
The speedup may be different with large textures and/or a different selection of mods. 
Joining the game may not be the only situation which is faster with the new `blit_with_alpha` function.
